fecgov/fec-proxy-redirect
# Set up new redirect app from 'app.cloud.gov' URLs to CDN URLs

Partial resolution for https://github.com/fecgov/fec-accounts/issues/313

See https://github.com/fecgov/fec-accounts/issues/289 for a diagram of how this will all work once it's complete

- As part of CMS restrictions in https://github.com/fecgov/fec-accounts/issues/289 we should have a redirect (302 temporary) from the `app.cloud.gov` routes to the CDN routes (ie. dev.fec.gov), in case the `app.cloud.gov` URLs come up in google search or are bookmarked or referenced anywhere else on the web (without this redirect app, the old app URLs will 404 after we implement https://github.com/fecgov/fec-accounts/issues/318)
- Borrowed heavily from https://github.com/fecgov/fec-proxy/ and simplified NGINX config for simple redirect app
- Set up CircleCI to deploy automatically (like recently merged https://github.com/fecgov/fec-proxy/pull/198)
- This repo will need to be included in releases starting with 13.8 (though I don't anticipate it changing much)

<img width="867" alt="Screen Shot 2020-11-18 at 4 34 23 PM" src="https://user-images.githubusercontent.com/31420082/99590911-ff132500-29bb-11eb-8593-7a5b9ede8c1b.png">


### Reviewers
At least two approvals, please

### How to test
- Make a copy of this branch and name it `release/<your_branch_name>` and/or add `|| ${branch} == "<your_branch_name>" ]]` to this line: https://github.com/fecgov/fec-proxy/blob/ed7387c8ac1ee68bf42b02610c9b1fcb8560406a/bin/cf_deploy.sh#L14
So it reads:
```
if [[ ${branch} == "develop" || ${branch} == "<your_branch_name>" ]]; then
```
- Push up your branch
- Your branch gets deployed to the `stage` space (or `dev` if you changed the dev rules above)
- Go to fec-stage-cms.app.cloud.gov or fec-dev-cms.app.cloud.gov (depending on which you chose)
- About 50% of the time, it will redirect to the CDN route (ie dev.fec.gov or stage.fec.gov) and the rest of the time, the application itself will load without redirecting.
- ❓ Why doesn't it work every time? This is because the CMS app and this redirect app are sharing this route (check this out with `cf apps` and see they share a route), so traffic will get routed to both apps. With this approach, we have a zero-downtime approach. If we remove the public route from the CMS (https://github.com/fecgov/fec-accounts/issues/318) before this redirect app is deployed, all the bookmarked and google search result URLs will 404.
